fml/raster_thread_merger_unittests.cc

```diff
@@ -623,5 +623,229 @@ TEST(RasterThreadMerger, SetMergeUnmergeCallback) {
+
+  latch1.Wait();
+  latch2.Wait();
+
+  fml::TaskQueueId qid1 = loop1->GetTaskRunner()->GetTaskQueueId();
+  fml::TaskQueueId qid2 = loop2->GetTaskRunner()->GetTaskQueueId();
+  const auto raster_thread_merger1_ =
+      fml::MakeRefCounted<fml::RasterThreadMerger>(qid1, qid2);
+  const auto raster_thread_merger2_ =
+      fml::MakeRefCounted<fml::RasterThreadMerger>(qid1, qid2);
+  const int kNumFramesMerged = 5;
+  ASSERT_FALSE(raster_thread_merger1_->IsMerged());
+  ASSERT_FALSE(raster_thread_merger2_->IsMerged());
+
+  // Merge using the second merger
+  raster_thread_merger2_->MergeWithLease(kNumFramesMerged);
+
+  // let there be one more turn till the leases expire.
+  for (int i = 0; i < kNumFramesMerged - 1; i++) {
+    // Check merge state using the first merger
+    ASSERT_TRUE(raster_thread_merger1_->IsMerged());
+    raster_thread_merger1_->DecrementLease();
+  }
+
+  // extend the lease once using the first merger
+  raster_thread_merger1_->ExtendLeaseTo(kNumFramesMerged);
+
+  // we will NOT last for 1 extra turn, we just set it.
+  for (int i = 0; i < kNumFramesMerged; i++) {
+    // Check merge state using the second merger
+    ASSERT_TRUE(raster_thread_merger2_->IsMerged());
+    raster_thread_merger2_->DecrementLease();
+  }
+
+  ASSERT_FALSE(raster_thread_merger1_->IsMerged());
+  ASSERT_FALSE(raster_thread_merger2_->IsMerged());
+
+  term1.Signal();
+  term2.Signal();
+  thread1.join();
+  thread2.join();
+}
+
+TEST(RasterThreadMerger, TheLastCallerOfMultipleMergersCanUnmergeNow) {
+  fml::MessageLoop* loop1 = nullptr;
+  fml::AutoResetWaitableEvent latch1;
+  fml::AutoResetWaitableEvent term1;
+  std::thread thread1([&loop1, &latch1, &term1]() {
+    fml::MessageLoop::EnsureInitializedForCurrentThread();
+    loop1 = &fml::MessageLoop::GetCurrent();
+    latch1.Signal();
+    term1.Wait();
+  });
+
+  fml::MessageLoop* loop2 = nullptr;
+  fml::AutoResetWaitableEvent latch2;
+  fml::AutoResetWaitableEvent term2;
+  std::thread thread2([&loop2, &latch2, &term2]() {
+    fml::MessageLoop::EnsureInitializedForCurrentThread();
+    loop2 = &fml::MessageLoop::GetCurrent();
+    latch2.Signal();
+    term2.Wait();
+  });
+
+  latch1.Wait();
+  latch2.Wait();
+
+  fml::TaskQueueId qid1 = loop1->GetTaskRunner()->GetTaskQueueId();
+  fml::TaskQueueId qid2 = loop2->GetTaskRunner()->GetTaskQueueId();
+  const auto raster_thread_merger1_ =
+      fml::MakeRefCounted<fml::RasterThreadMerger>(qid1, qid2);
+  const auto raster_thread_merger2_ =
+      fml::MakeRefCounted<fml::RasterThreadMerger>(qid1, qid2);
+  const int kNumFramesMerged = 5;
+  ASSERT_FALSE(raster_thread_merger1_->IsMerged());
+  ASSERT_FALSE(raster_thread_merger2_->IsMerged());
+
+  // Recording mergers themselves is needed.
+  raster_thread_merger1_->RecordMergeCaller();
+  raster_thread_merger2_->RecordMergeCaller();
+  // Merge using the mergers
+  raster_thread_merger1_->MergeWithLease(kNumFramesMerged);
+  raster_thread_merger2_->MergeWithLease(kNumFramesMerged);
```

Comment:
Can you add asserts for IsMerged between the RecordMergeCaller and MergeWithLease calls, and between the 2 MergeWithLease calls?

Reply:
Done~